### DISCUSSION TITLE
[RFC] 002 - Pref Mobile 移动端 FOUC

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#252] [RFCs]
## 问题

由于我们在移动端和桌面端采用两套完全不同的布局，因此需要通过 JS 来判断布局组件的渲染。

由于当前判断 mobile  采用了 客户端 JS 判断，会导致移动端的 [FOUC](https://zhuanlan.zhihu.com/p/90172207)。

https://github.com/lobehub/lobe-chat/assets/28616219/f277c108-32d3-49c8-827e-fc83b69318eb

此前在 SSG 下此问题基本无解。 

在 SSR 下可以采用 cookie 或者 ua 的方式做判断优化首屏，但仍然会引入两套路由布局组件，导致 js bundle 相对较大。

在切换到app router后，我们可以尝试 rsc 从根本上解决该问题。

## 思路

将mobile判断从 `const { mobile } = useResponsive(); ` 的 client 端判断，换成服务端判断。

目前看到比较合适的库是 [ua-parser-js](https://github.com/faisalman/ua-parser-js) ，体验 demo: https://uaparser.js.org/

在服务端中使用的方式：

```js
const result = UAParser(request.headers); // parse only "user-agent" header
```

## 改造代码

以 [welcome](https://github.com/lobehub/lobe-chat/blob/99f72b805bdcaaa562191de84ff39ac29ad56c09/src/app/welcome/index.tsx#L25) 为例 :

当前的实现为
```tsx
'use client';

const Welcome = () => {
  const { mobile } = useResponsive();

  return mobile ? (
    <Mobile>
      <Banner />
    </Mobile>
  ) : (
    <Layout>
      <Banner />
      <Footer />
    </Layout>
  );
};

export default Welcome;
```

改造后：
```tsx
// layout.tsx
import { PropsWithChildren } from 'react';

import { isMobileDevice } from '@/utils/responsive';

import Desktop from './DesktopLayout';
import Mobile from './MobileLayout';
import Footer from './features/Footer';

const Layout = ({ children }: PropsWithChildren) => {
  const mobile = isMobileDevice();

  return mobile ? (
    <Mobile>{children}</Mobile>
  ) : (
    <Desktop>
      {children}
      <Footer />
    </Desktop>
  );
};

export default Layout;
```

## 待明确点

1. page和 layout 怎么区分的最佳实践？
2. 相应的性能和尺寸如何评估？
3. Desktop 下仍然存在响应式，所以desktop 的jsbundle尺寸小不了？且如果 pc 上浏览器是小尺寸的话，可能仍然有 FOUC 的问题。

**Comment by arvinxx** [comment]
- [x] welcome https://github.com/lobehub/lobe-chat/pull/257
- [x] chat https://github.com/lobehub/lobe-chat/pull/264
- [x] settings https://github.com/lobehub/lobe-chat/pull/279
- [x] market https://github.com/lobehub/lobe-chat/pull/278

**Comment by arvinxx** [comment]
chat 路由改造效果：

<img width="1166" alt="image" src="https://github.com/lobehub/lobe-chat/assets/28616219/16b849ed-4a1f-4720-9c32-b738b4d330d8">

https://twitter.com/arvin17x/status/1710860673499177393?s=61&t=ZkWAwVnwODUo0Bew3JG2Kg

**Comment by canisminor1990** [comment]
# Lighthouse Reports https://github.com/lobehub/lobe-chat/pull/279

#### TOC

- [Welcome Page](#welcome-page)
- [Chat Page](#chat-page)
- [Market Page](#market-page)
- [Settings Page](#settings-page)

## Welcome Page

> **Info**\
> <https://chat-preview.lobehub.com/welcome>

|                     Desktop                     |                     Mobile                     |
| :---------------------------------------------: | :--------------------------------------------: |
|              ![][welcome-desktop]               |              ![][welcome-mobile]               |
| [⚡️ Lighthouse Report][welcome-desktop-report] | [⚡️ Lighthouse Report][welcome-mobile-report] |

## Chat Page

> **Info**\
> <https://chat-preview.lobehub.com/chat>

|                   Desktop                    |                   Mobile                    |
| :------------------------------------------: | :-----------------------------------------: |
|              ![][chat-desktop]               |              ![][chat-mobile]               |
| [⚡️ Lighthouse Report][chat-desktop-report] | [⚡️ Lighthouse Report][chat-mobile-report] |

## Market Page

> **Info**\
> <https://chat-preview.lobehub.com/market>

|                    Desktop                     |                    Mobile                     |
| :--------------------------------------------: | :-------------------------------------------: |
|              ![][market-desktop]               |              ![][market-mobile]               |
| [⚡️ Lighthouse Report][market-desktop-report] | [⚡️ Lighthouse Report][market-mobile-report] |

## Settings Page

> **Info**\
> <https://chat-preview.lobehub.com/settings>

|                     Desktop                      |                     Mobile                      |
| :----------------------------------------------: | :---------------------------------------------: |
|              ![][settings-desktop]               |              ![][settings-mobile]               |
| [⚡️ Lighthouse Report][settings-desktop-report] | [⚡️ Lighthouse Report][settings-mobile-report] |

[chat-desktop]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/chat/desktop/pagespeed.svg
[chat-desktop-report]: https://lobehub.github.io/lobe-chat/lighthouse/chat/desktop/chat_preview_lobehub_com_chat.html
[chat-mobile]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/chat/mobile/pagespeed.svg
[chat-mobile-report]: https://lobehub.github.io/lobe-chat/lighthouse/chat/mobile/chat_preview_lobehub_com_chat.html
[market-desktop]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/market/desktop/pagespeed.svg
[market-desktop-report]: https://lobehub.github.io/lobe-chat/lighthouse/market/desktop/chat_preview_lobehub_com_market.html
[market-mobile]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/market/mobile/pagespeed.svg
[market-mobile-report]: https://lobehub.github.io/lobe-chat/lighthouse/market/mobile/chat_preview_lobehub_com_market.html
[settings-desktop]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/settings/desktop/pagespeed.svg
[settings-desktop-report]: https://lobehub.github.io/lobe-chat/lighthouse/settings/desktop/chat_preview_lobehub_com_settings.html
[settings-mobile]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/settings/mobile/pagespeed.svg
[settings-mobile-report]: https://lobehub.github.io/lobe-chat/lighthouse/settings/mobile/chat_preview_lobehub_com_settings.html
[welcome-desktop]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/welcome/desktop/pagespeed.svg
[welcome-desktop-report]: https://lobehub.github.io/lobe-chat/lighthouse/welcome/desktop/chat_preview_lobehub_com_welcome.html
[welcome-mobile]: https://raw.githubusercontent.com/lobehub/lobe-chat/lighthouse/lighthouse/welcome/mobile/pagespeed.svg
[welcome-mobile-report]: https://lobehub.github.io/lobe-chat/lighthouse/welcome/mobile/chat_preview_lobehub_com_welcome.html
